pkg/controllers/provisioning/scheduling/topology_test.go

```diff
@@ -1125,6 +1126,544 @@ var _ = Describe("Topology", func() {
+						},
+					},
+					TopologySpreadConstraints: topology,
+				}, 5)...,
+			)
+			// we're aware of three domains, but can only schedule a single pod to the domain where we're allowed
+			// to create nodes
+			ExpectSkew(ctx, env.Client, "default", &topology[0]).To(ConsistOf(1))
+		})
+		It("should balance pods across a label (NodeTaintsPolicy=honor)", func() {
+			const spreadLabel = "fake-label"
+			nodePool.Spec.Template.Labels = map[string]string{
+				spreadLabel: "baz",
+			}
+			node1 := test.Node(test.NodeOptions{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						spreadLabel: "foo",
+					},
+				},
+				Taints: []corev1.Taint{
+					{
+						Key:    "taintname",
+						Value:  "taintvalue",
+						Effect: corev1.TaintEffectNoSchedule,
+					},
+				},
+				Allocatable: corev1.ResourceList{
+					corev1.ResourceCPU: resource.MustParse("100m"),
+				}})
+
+			node2 := test.Node(test.NodeOptions{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						spreadLabel: "bar",
+					},
+				},
+				Taints: []corev1.Taint{
+					{
+						Key:    "taintname",
+						Value:  "taintvalue",
+						Effect: corev1.TaintEffectNoSchedule,
+					},
+				},
+				Allocatable: corev1.ResourceList{
+					corev1.ResourceCPU: resource.MustParse("100m"),
+				}})
+			ExpectApplied(ctx, env.Client, nodePool, node1, node2)
+
+			ExpectReconcileSucceeded(ctx, nodeStateController, client.ObjectKeyFromObject(node1))
+			ExpectReconcileSucceeded(ctx, nodeStateController, client.ObjectKeyFromObject(node2))
+
+			// since two of the nodes are tainted, Karpenter should not consider them for topology domain discovery
+			// purposes
+			ExpectProvisioned(ctx, env.Client, cluster, cloudProvider, prov)
+			honor := corev1.NodeInclusionPolicyHonor
+			topology := []corev1.TopologySpreadConstraint{{
+				TopologyKey:       spreadLabel,
+				WhenUnsatisfiable: corev1.DoNotSchedule,
+				LabelSelector:     &metav1.LabelSelector{MatchLabels: labels},
+				MaxSkew:           1,
+				NodeTaintsPolicy:  &honor,
+			}}
+
+			ExpectApplied(ctx, env.Client, nodePool)
+			ExpectProvisioned(ctx, env.Client, cluster, cloudProvider, prov,
+				test.UnschedulablePods(test.PodOptions{
+					ObjectMeta: metav1.ObjectMeta{Labels: labels},
+					ResourceRequirements: corev1.ResourceRequirements{
+						Requests: corev1.ResourceList{
+							corev1.ResourceCPU: resource.MustParse("1"),
+						},
+					},
+					TopologySpreadConstraints: topology,
+				}, 5)...,
+			)
+			// and should schedule all of the pods on the same node
+			ExpectSkew(ctx, env.Client, "default", &topology[0]).To(ConsistOf(5))
+		})
+		It("should balance pods across a label when discovered from the provisioner (NodeTaintsPolicy=ignore)", func() {
+			const spreadLabel = "fake-label"
+			const taintKey = "taint-key"
+			nodePool.Spec.Template.Spec.Requirements = append(nodePool.Spec.Template.Spec.Requirements, v1.NodeSelectorRequirementWithMinValues{
```

Comment:
For the tests that are validating schedulability of a pod against a nodepool or node, it might be worth it to add taints and tolerations for the nodepool/node that we are able to schedule to as well so that we ensure that we are exercising the tolerates checks that we have in the topology code

Reply:
I could also see an argument that this is tested below but it probably doesn't hurt -- I could go either way

Reply:
I've updated the test below this one to exercise this, and it did actually catch a miss in the check for taints / tolerations in `topology.Record`.

Reply:
Nice! Good to hear!